packages/frontend/src/hooks/useChartConfig.tsx

```diff
@@ -163,29 +176,58 @@ export const useChartConfig = (
               )
             : queryResults.data.rows;
         const { groupDimension } = seriesLayout;
-        const series = groupDimension
-            ? Array.from(
-                  new Set(queryResults.data.rows.map((r) => r[groupDimension])),
-              )
-            : seriesLayout.yMetrics;
+        let series: string[];
+        const eChartDimensions = [
+            {
+                name: seriesLayout.xDimension,
+                displayName: friendlyName(seriesLayout.xDimension),
+            },
+        ];
+        // add more chart dimensions
+        if (groupDimension) {
+            series = Array.from(
+                new Set(queryResults.data.rows.map((r) => r[groupDimension])),
+            );
+            const dimensionFormatter = getDimensionFormatter(
+                dimensions.find(
+                    (value) => getFieldId(value) === groupDimension,
+                ) as CompiledDimension,
+            );
+            // Convert to dictionary to get unique set.
+            const groupChartDimensions: {
+                name: string;
+                displayName: string;
+            }[] = Object.values(
+                queryResults.data.rows.reduce(
+                    (coll, row) => ({
+                        ...coll,
+                        [row[groupDimension as string]]: {
+                            name: row[groupDimension as string],
+                            displayName:
+                                dimensionFormatter({
+                                    value: row[groupDimension],
+                                }) ?? row[groupDimension],
+                        },
+                    }),
+                    {},
+                ),
+            );
+            eChartDimensions.push(...groupChartDimensions);
+        } else {
+            series = seriesLayout.yMetrics;
+            const yMetricChartDimensions = series.map((s) => ({
+                name: s,
+                displayName: friendlyName(s),
+            }));
+            eChartDimensions.push(...yMetricChartDimensions);
+        }
```

Comment:
Seems like we could improve this further by: 
- when declaring `dimensionFormatter` we are forcing the type `CompiledDimension` inside which can cause issues in the future. We can make the code safer by asserting there is group compiled dimension object before creating the dimension formatter
- for `groupChartDimensions` we can reuse the existing typing `ChartConfig['eChartDimensions']` so it's easier to update the typing in the future
- for `groupChartDimensions` we don't need to use the `queryResults` again since we already have the unique keys in `series`

This is my proposal, let me know what you think. 

```
        const groupCompiledDimension = dimensions.find(
            (value) => getFieldId(value) === groupDimension,
        );

        let series: string[];
        const eChartDimensions: ChartConfig['eChartDimensions'] = [
            {
                name: seriesLayout.xDimension,
                displayName: friendlyName(seriesLayout.xDimension),
            },
        ];

        if (groupCompiledDimension && groupDimension) {
            const dimensionFormatter = getDimensionFormatter(
                groupCompiledDimension,
            );
            series = Array.from(
                new Set(queryResults.data.rows.map((r) => r[groupDimension])),
            );
            const groupChartDimensions: ChartConfig['eChartDimensions'] =
                series.map((s) => ({
                    name: s,
                    displayName: dimensionFormatter({
                        value: s,
                    }),
                }));
            eChartDimensions.push(...groupChartDimensions);
        } else {
            series = seriesLayout.yMetrics;
            const yMetricChartDimensions = series.map((s) => ({
                name: s,
                displayName: friendlyName(s),
            }));
            eChartDimensions.push(...yMetricChartDimensions);
        }
```

Reply:
### 

Thanks for pointing out to add the typing!. The current solution is quite closer to the one I wrote at first :smile_cat: before messing up thinking dimension formatter was dependent on rows.

I don't really like the idea of grouping `groupCompiledDimension` and `groupDimension` together. `groupCompiledDimension` doesn't mean yMetric should be used. 
`groupCompiledDimension` null is actually an error, would prefer for it to throw an error and catch it early rather than passing it silently. Another alternate would be using optional

```typescript jsx
        let series: string[];
        const eChartDimensions : ChartConfig['eChartDimensions'] = [
            {
                name: seriesLayout.xDimension,
                displayName: friendlyName(seriesLayout.xDimension),
            },
        ];
        // add more chart dimensions
        if (groupDimension) {
            series = Array.from(
                new Set(queryResults.data.rows.map((r) => r[groupDimension])),
            );
            const groupCompiledDimension = dimensions.find(
                (value) => getFieldId(value) === groupDimension,
            );
            const dimensionFormatter = groupCompiledDimension ? getDimensionFormatter(groupCompiledDimension) : null;
            // Convert to dictionary to get unique set.
            const groupChartDimensions: {
                name: string;
                displayName: string;
            }[] = series.map((s) => ({
                            name: s,
                            displayName:
                                dimensionFormatter?.({
                                    value: s,
                                }) ?? friendlyName(s)
            }));
            eChartDimensions.push(...groupChartDimensions);
        } else {
            series = seriesLayout.yMetrics;
            const yMetricChartDimensions = series.map((s) => ({
                name: s,
                displayName: friendlyName(s),
            }));
            eChartDimensions.push(...yMetricChartDimensions);
        }
```

Reply:
> I don't really like the idea of grouping groupCompiledDimension and groupDimension together. groupCompiledDimension doesn't mean yMetric should be used.

Good point! I think we can go with the approach you mentioned instead of throwing an error.